go/network/frontend.go

```diff
@@ -0,0 +1,392 @@
+	}
+}
+
+func (p *Frontend) serveConn(ctx context.Context, raw net.Conn) {
+	var conn = &frontendConn{
+		id:  reflect.ValueOf(raw).Pointer(),
+		ctx: ctx,
+		raw: raw,
+		tls: tls.Server(raw, p.tlsConfig),
+	}
+
+	// Push `conn` onto the map of current handshakes.
+	p.handshakeMu.Lock()
+	p.handshake[conn.id] = conn
+	p.handshakeMu.Unlock()
+
+	// The TLS handshake machinery will next call into getTLSConfigForClient().
+	var err = conn.tls.HandshakeContext(conn.ctx)
+
+	// Clear `conn` from the map of current handshakes.
+	p.handshakeMu.Lock()
+	delete(p.handshake, conn.id)
+	p.handshakeMu.Unlock()
+
+	if err != nil {
+		handshakeCounter.WithLabelValues(err.Error()).Inc() // `err` is low-variance.
+		p.serveConnErr(conn.raw, 421, "This service may only be accessed using TLS, such as through an https:// URL.\n")
+		return
+	}
+	if conn.pass {
+		handshakeCounter.WithLabelValues("OKPass").Inc()
+		p.fwdCh <- conn.tls // Connection is not for us.
+		return
+	}
+
+	if conn.sniErr != nil {
+		handshakeCounter.WithLabelValues("ErrSNI").Inc()
+		p.serveConnErr(conn.tls, 404, fmt.Sprintf("Failed to match the connection to a task:\n\t%s\n", conn.sniErr))
+	} else if conn.dialErr != nil {
+		handshakeCounter.WithLabelValues("ErrDial").Inc()
+		p.serveConnErr(conn.tls, 503, fmt.Sprintf("Failed to connect to a task shard:\n\t%s\n", conn.dialErr))
+	} else if conn.dialed != nil {
+		handshakeCounter.WithLabelValues("OkTCP").Inc()
+		p.serveConnTCP(conn)
+	} else {
+		handshakeCounter.WithLabelValues("OkHTTP").Inc()
+		p.serveConnHTTP(conn)
+	}
+}
+
+func (p *Frontend) getTLSConfigForClient(hello *tls.ClientHelloInfo) (*tls.Config, error) {
+	p.handshakeMu.Lock()
+	var conn = p.handshake[reflect.ValueOf(hello.Conn).Pointer()]
+	p.handshakeMu.Unlock()
+
+	// Exact match of the FQDN or a parent domain means it's not for us.
+	if slices.Contains(p.domains, hello.ServerName) {
+		conn.pass = true
+		return nil, nil
+	}
+
+	var ok bool
+	var target, service, _ = strings.Cut(hello.ServerName, ".")
+
+	// This block parses the SNI `target` and matches it to shard configuration.
+	if !slices.Contains(p.domains, service) {
+		conn.sniErr = fmt.Errorf("TLS ServerName %s is an invalid domain", hello.ServerName)
+	} else if conn.parsed, conn.sniErr = parseSNI(target); conn.sniErr != nil {
+		// No need to wrap error.
+	} else if conn.resolved, ok = p.sniCache.Get(conn.parsed); !ok {
+		// We didn't hit cache while resolving the parsed SNI.
+		// We must fetch matching shard specs to inspect their shard ID prefix and port config.
+		var shards []pc.ListResponse_Shard
+		shards, conn.sniErr = listShards(hello.Context(), p.shardClient, conn.parsed, "")
+
+		if conn.sniErr != nil {
+			conn.sniErr = fmt.Errorf("fetching shards: %w", conn.sniErr)
+		} else if len(shards) == 0 {
+			conn.sniErr = errors.New("the requested subdomain does not match a known task and port combination")
+		} else {
+			conn.resolved = newResolvedSNI(conn.parsed, &shards[0].Spec)
```

Comment:
This seems fine for now, but it might not be ideal to always pick the first shard in cases where there's multiple shards and the hostname doesn't specify a key+rclock. It seems better to pick on randomly in order to distribute the load. I don't feel super confident that random selection would be ideal either, but wanted to raise the idea since splitting `source-http-ingest` shards had been on my mind lately.

Reply:
This is not used to dial a shard, it's only used to fetch task metadata, and any shard will do -- we're just plucking out the shard ID , port protocol, and whether it's public.

We list shards again and randomize them when picking a primary to dial.